hypervisor/src/kvm.rs

```diff
@@ -0,0 +1,273 @@
+// Copyright © 2019 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Copyright © 2020, Microsft  Corporation
+//
+#[cfg(target_arch = "x86_64")]
+use crate::cpuidpatch::{self, patch_cpuid, CpuidPatch, CpuidReg};
+
+use crate::bindings::*;
+use crate::{GenVcpuFd, GenVmFd, Hypervisor};
+#[cfg(target_arch = "x86_64")]
+use devices::ioapic;
+#[cfg(target_arch = "x86_64")]
+use kvm_bindings::{kvm_enable_cap, CpuId, KVM_CAP_SPLIT_IRQCHIP};
+use kvm_ioctls::{Cap, DeviceFd, IoEventAddress, Kvm, NoDatamatch, VcpuFd, VmFd};
+use std::result;
+use std::sync::Arc;
+use vm_memory::{Address, GuestAddress};
+
+use vmm_sys_util::eventfd::EventFd;
+pub const KVM_TSS_ADDRESS: GuestAddress = GuestAddress(0xfffb_d000);
+extern crate linux_loader;
+
+pub struct KvmVmFd {
+    fd: Arc<VmFd>,
+    #[cfg(target_arch = "x86_64")]
+    cpuid: CpuId,
+}
+impl GenVmFd for KvmVmFd {
+    #[cfg(target_arch = "x86_64")]
+    fn set_tss_address(&self, offset: usize) -> ResultOps<()> {
+        self.fd.set_tss_address(offset)
+    }
+    fn create_irq_chip(&self) -> ResultOps<()> {
+        self.fd.create_irq_chip()
+    }
+    fn register_irqfd(&self, fd: &EventFd, gsi: u32) -> ResultOps<()> {
+        self.fd.register_irqfd(fd, gsi)
+    }
+    fn unregister_irqfd(&self, fd: &EventFd, gsi: u32) -> ResultOps<()> {
+        self.fd.unregister_irqfd(fd, gsi)
+    }
+    fn create_vcpu(&self, id: u8) -> ResultOps<Arc<dyn GenVcpuFd>> {
+        let vc = self.fd.create_vcpu(id).expect("new VcpuFd failed");
+        let vcpu = KvmVcpuId { fd: vc };
+        Ok(Arc::new(vcpu))
+    }
+    fn register_ioevent(
+        &self,
+        fd: &EventFd,
+        addr: &IoEventAddress,
+        datamatch: Option<u64>,
+    ) -> ResultOps<()> {
+        if let Some(kvm_datamatch) = datamatch {
+            self.fd.register_ioevent(fd, addr, kvm_datamatch)
+        } else {
+            self.fd.register_ioevent(fd, addr, NoDatamatch)
+        }
+    }
+    fn unregister_ioevent(&self, fd: &EventFd, addr: &IoEventAddress) -> ResultOps<()> {
+        self.fd.unregister_ioevent(fd, addr)
+    }
+    fn set_gsi_routing(&self, irq_routing: &IrqRouting) -> ResultOps<()> {
+        self.fd.set_gsi_routing(irq_routing)
+    }
+    fn set_user_memory_region(&self, user_memory_region: MemoryRegion) -> ResultOps<()> {
+        unsafe { self.fd.set_user_memory_region(user_memory_region) }
```

Comment:
For unsafe blocks, we should add a comment describing why we think it is still safe.